SBT build improvements

retronym opened the pull request:
Usability improvements: aliases and tab completion for partest.

Review by @szeiger

retronym committed:
Convenient aliases for the SBT build

The idea here is to avoid the need for switching
between the SBT shell and Bash.

  - Add aliases for partest, scala{,c,p,doc}
  - Change working directory for these forked processes
    to the root project base directory, rather than the
    subprojects base directory. This lets us use relative
    paths to files in a more familar way.
  - Don't log the output of forked processes with the
    `[info] ` prefix, rather pass it through directly
    to stdout.

Demo:

```
> partest --terse test/files/pos/t6231.scala
[info] Packaging /Users/jason/code/scala2/build-sbt/pack/lib/scala-partest-javaagent.jar ...
[info] Done packaging.
Selected 1 tests drawn from specified tests

.

[info] Passed: Total 1, Failed 0, Errors 0, Passed 1
[success] Total time: 3 s, completed 05/02/2016 12:44:19 PM
> scala sandbox/test.scala
[info] Running scala.tools.nsc.MainGenericRunner -usejavacp sandbox/test.scala
Hello, World!
[success] Total time: 4 s, completed 05/02/2016 12:45:08 PM
> scalac sandbox/test.scala
[info] Running scala.tools.nsc.Main -usejavacp sandbox/test.scala
[success] Total time: 3 s, completed 05/02/2016 12:45:15 PM
> scala Test
[info] Running scala.tools.nsc.MainGenericRunner -usejavacp Test
Hello, World!
[success] Total time: 1 s, completed 05/02/2016 12:45:20 PM
> scala
[info] Running scala.tools.nsc.MainGenericRunner -usejavacp
Welcome to Scala 2.11.8-20160204-090931-42525ec (Java HotSpot(TM) 64-Bit Server VM, Java 1.8.0_71).
Type in expressions for evaluation. Or try :help.

scala> 1.to
to   toBinaryString   toByte   toChar   toDegrees   toDouble   toFloat   toHexString   toInt   toLong   toOctalString   toRadians   toShort   toString

scala> 1.toString
res0: String = 1

scala> :quit
[success] Total time: 8 s, completed 05/02/2016 12:45:48 PM
>
```

retronym committed:
Add tab completion to the partest command

We can complete partest options (I've excluded some that
aren't relevant in SBT), as well as test file names.

If `--srcpath scaladoc` is included, completion of test
paths will be based on `test/scaladoc` rather than
the default `test/files`. Note that the `--srcpath` option
is currently broken via scala partest interface, this
change to scala-partest is needed to make it work:

   https://github.com/scala/scala-partest/pull/49

I've also hijacked the `--grep` option with logic
in the SBT command itself, rather than passing this
to `partest`. Just like `./bin/partest-ack`, this looks
for either test file names or regex matches within
the contents of test, check, or flag files.

I tried for some time to make the tab completion of
thousands of filenames more user friendly, but wasn't
able to get something working. Ideally, it should
only suggest to `test/files/{pos, neg, ...}` on the
first <TAB>, and then offer files on another TAB.
Files should also be offered if a full directory
has been entered. Hopefully a SBT parser guru will
step in and add some polish here.

retronym commented:
/cc @dwijnand for any tips on making the parser less hacky...

dwijnand commented:
I don't know sbt's parsing infrastructure, but I'd look at how scripted is coded and/or ask @eed3si9n.

I'll try and have a look at this this weekend though.

retronym commented:
Thanks. Don't go too far down the rabbit hole in my behalf... :)

I took a look at the scripted parser as inspiration but didn't figure it out.

retronym commented:
Just added a commit to mimic what the IntelliJ SBT import would do (were it able to handle our build), so that we can get IDE smarts when editing the build definition.

retronym committed:
Enable IntellIJ smarts while editing the build

Even though our build can't be imported into IntelliJ
automatically, we don't need to live in the dark ages
of red squigglies when editing build.sbt.

This commit hand-crafts the module defintion for the
project, which has SBT and its dependendencies on the
classpath.

A screenshot of the code assist is worth the thousand
expletives I uttered while writing the pictured code
without this facility:

  https://www.dropbox.com/s/6mxv0iwxkhvnor7/Screenshot%202016-02-07%2022.09.12.png?dl=0

To download the sources for the SBT JARs that are referenced
herein, run

    sbt> reload plugins
    sbt> updateClassifiers

retronym commented:
> Thanks. Don't go too far down the rabbit hole in my behalf... :)

Or, if rabbit holes are your thing, you might like to write an option parser for the `{scalac, scala, scaladoc}` commands in the same manner as https://github.com/scala/scala-tool-support/blob/master/bash-completion/src/main/scala/BashCompletion.scala

First version would be a static parser, based on the current set of options that enumerate by running the code above against 2.11.7. The extra credit version would fire up the just-built compiler to get the options-du-jour. That would need to be done via reflection, of course (in the same way that `runMain` works).

dwijnand commented:
Yeah I've been looking at this since yesterday.. and I'm throwing in the towel, sorry. It's an area I've just skim read in the docs and in the sources, not anything I'm well-versed on.

retronym committed:
Add SBT tab completion for scala{,c,doc}

Also refactor the partest parser to use the improved tab
completion for test paths.

retronym commented:
I've just pushed an additional commit with my first attempt at the Scala option parser.

szeiger commented on build.sbt:
```diff
@@ -200,7 +200,18 @@ lazy val commonSettings = clearSourceAndResourceDirectories ++ publishSettings +
+  packageOptions in Compile in packageSrc := Seq.empty,
+
+  // Lets us CTRL-C partest without exiting SBT entirely
+  cancelable in Global := true,
```
You may want to reference https://github.com/sbt/sbt/issues/1442 here. Seeing the brute force `in Global` in a subproject definition made me wonder why you need it, leading to said sbt ticket.

szeiger commented:
Awesome improvements. I can't really offer any advise on the sbt parser combinator code. I learned the basics only a few weeks ago when working on the Scala build (and then abandoned the solution that would have required me to write parsers due to other complications).

retronym commented:
A neat trick built on top of this is to use continuous execution in SBT like:

```
// local.sbt (.gitignored)
watchSources ++= (((baseDirectory in ThisBuild).value / "sandbox") ** ("*.java" | "*.scala")).get
```

```
> ~ ; scalac sandbox/test.scala; scala Test
```

That will retrigger the continuous execution when you change the compiler or when you change the `test.scala` file.

eed3si9n commented:
This reminds me of https://github.com/mentat-labs/sbt-nsa, which tries to strongly type scalac options.

retronym commented:
> Could be risky if you have whitespace or escaped characters in the command line and partestParser doesn't agree with sbt's own parser on tokenization.

Yep, that's definitely a hole. Passing a single command string from our custom task to SBT's "remainingCommands" is super easy, but it would be more principled for us to delegate to the internals of `runMain` and `testOnly` programatically. However it seems we might have to copy/paste a bit of code from those tasks to make that happen.

```
  def runMainTask(classpath: Initialize[Task[Classpath]], scalaRun: Initialize[Task[ScalaRun]]): Initialize[InputTask[Unit]] =
    {
      import DefaultParsers._
      val parser = loadForParser(discoveredMainClasses)((s, names) => runMainParser(s, names getOrElse Nil))
      Def.inputTask {
        val (mainClass, args) = parser.parsed
        toError(scalaRun.value.run(mainClass, data(classpath.value), args, streams.value.log))
      }
    }
```

```
def inputTests(key: InputKey[_]): Initialize[InputTask[Unit]] = inputTests0.mapReferenced(Def.mapScope(_ in key.key))
  private[this] lazy val inputTests0: Initialize[InputTask[Unit]] =
    {
      val parser = loadForParser(definedTestNames)((s, i) => testOnlyParser(s, i getOrElse Nil))
      Def.inputTaskDyn {
        val (selected, frameworkOptions) = parser.parsed
        val s = streams.value
        val filter = testFilter.value
        val config = testExecution.value

        implicit val display = Project.showContextKey(state.value)
        val modifiedOpts = Tests.Filters(filter(selected)) +: Tests.Argument(frameworkOptions: _*) +: config.options
        val newConfig = config.copy(options = modifiedOpts)
        val output = allTestGroupsTask(s, loadedTestFrameworks.value, testLoader.value, testGrouping.value, newConfig, fullClasspath.value, javaHome.value, testForkedParallel.value)
        val taskName = display(resolvedScoped.value)
        val trl = testResultLogger.value
        val processed = output.map(out => trl.run(s.log, out, taskName))
        Def.value(processed)
      }
    }
```

szeiger commented:
Ah, and half the stuff you'd call from the duplicated code is probably private.

SethTisue commented:
further improvements are always possible, but there isn't any reason not to go ahead and merge this, is there?

retronym commented:
I think this is ready to go. We might reconsider the scoping of the `baseDirectory in {Compile, Test}` setting (e.g. I could temporarily set that in the custom commands, rather than set it in the build proper), but let's wait until evidence that something breaks.

Separately, we should release a new version of partest (I merged my change to honour `--srcpath`) and update `versions.properties`), and then merge all of this to 2.12.x.
